traits/ctraits.c

```diff
@@ -2190,76 +2190,55 @@ call_notifiers(
     // Do nothing if the user has explicitly requested no traits notifications
     // to be sent.
     if ((obj->flags & HASTRAITS_NO_NOTIFY)) {
-        goto exit2;
+        goto exit;
     }
 
     if (tnotifiers != NULL) {
-        n = PyList_GET_SIZE(tnotifiers);
-        temp = NULL;
-        if (n > 1) {
-            temp = PyList_New(n);
-            if (temp == NULL) {
-                rc = -1;
-                goto exit2;
-            }
-            for (i = 0; i < n; i++) {
-                item = PyList_GET_ITEM(tnotifiers, i);
-                PyList_SET_ITEM(temp, i, item);
-                Py_INCREF(item);
-            }
-            tnotifiers = (PyListObject *)temp;
-        }
-        for (i = 0; i < n; i++) {
-            if (new_value_has_traits
-                && (((has_traits_object *)new_value)->flags
-                    & HASTRAITS_VETO_NOTIFY)) {
-                goto exit;
-            }
-            result = PyObject_Call(PyList_GET_ITEM(tnotifiers, i), args, NULL);
-            if (result == NULL) {
-                rc = -1;
-                goto exit;
-            }
-            Py_DECREF(result);
-        }
-        Py_XDECREF(temp);
+        t_len = PyList_GET_SIZE(tnotifiers);
+    } else {
+        t_len = 0;
     }
-
-    temp = NULL;
     if (onotifiers != NULL) {
-        n = PyList_GET_SIZE(onotifiers);
-        if (n > 1) {
-            temp = PyList_New(n);
-            if (temp == NULL) {
-                rc = -1;
-                goto exit2;
-            }
-            for (i = 0; i < n; i++) {
-                item = PyList_GET_ITEM(onotifiers, i);
-                PyList_SET_ITEM(temp, i, item);
-                Py_INCREF(item);
-            }
-            onotifiers = (PyListObject *)temp;
+        o_len = PyList_GET_SIZE(onotifiers);
+    } else {
+        o_len = 0;
+    }
+
+    // Concatenating trait notifiers and object notifiers.
+    // Notifier lists are copied in order to prevent run-time modifications.
```

Comment:
Is it still possible to have this runtime modification i.e. the list is modified by another thread while we loop over the list here?

Reply:
I think we're safe here because of the GIL: there's no way that I can see for another Python thread to intervene between the time that we check the lengths of `onotifiers` and `tnotifiers` and the time that we finish constructing the list of all notifiers.

Reply:
:+1: Thanks. I was overthinking the acquiring and releasing of GIL.